### PR TITLE
checkers: add new checker stringsCompare

### PR DESCRIPTION
**Pull request by peakle**
pr for: https://github.com/go-critic/go-critic/issues/1167

### CODE REVIEW

File: checkers/rules/rules.go [hunk 1]
```diff
@@ -736,3 +736,34 @@ func dynamicFmtString(m dsl.Matcher) {
...
 		Report(`use errors.New($f($*args)) or fmt.Errorf("%s", $f($*args)) instead`)
 }
+
+//doc:summary Detects strings.Compare usage
+//doc:tags    performance experimental
+//doc:before  strings.Compare(x, y)
+//doc:after   x < y
+func stringsCompare(m dsl.Matcher) {
+	m.Match(`strings.Compare($s1, $s2) == 0`).
+		Report(`don't use strings.Compare on hot path, change it to built-in operators`).
```
Comment (quasilyte) [inline checkers/rules/rules.go:746]: This warning message is not helpful.
Suggest template is more explanatory.
In general, the warning message should tell how to fix the situation.
In other words, it should tell what to do instead of saying what not to do.

File: checkers/rules/rules.go [hunk 2]
```diff
@@ -736,3 +736,34 @@ func dynamicFmtString(m dsl.Matcher) {
 		Suggest("errors.New($f($*args))").
 		Report(`use errors.New($f($*args)) or fmt.Errorf("%s", $f($*args)) instead`)
 }
+
+//doc:summary Detects strings.Compare usage
+//doc:tags    performance experimental
```
Comment (quasilyte) [inline checkers/rules/rules.go:741]: It's up to a debate whether this is a performance-first issue.
I would say most patterns would fall to style category, like using `==` instead of `strings.Compare`.
Other, less obvious patterns, can go to a separate, performance-tagged checker.

File: checkers/rules/rules.go [hunk 3]
```diff
@@ -736,3 +736,34 @@ func dynamicFmtString(m dsl.Matcher) {
...
+		Suggest(`$s1 < $s2`).
+		At(m["s1"])
+
+	m.Match(`strings.Compare($s1, $s2) > 0`,
+		`strings.Compare($s1, $s2) == 1`).
+		Report(`don't use strings.Compare on hot path, change it to built-in operators`).
+		Suggest(`$s1 > $s2`).
+		At(m["s1"])
+
+	m.Match(`$_ = strings.Compare($s1, $_)`,
```
Comment (quasilyte) [inline checkers/rules/rules.go:762]: This is too general and will mark any usages of `strings.Compare`.
Giving false positives is bad.

**Comment by quasilyte** [comment]
I would do this in 2 steps:
1. suggest simpler operations for the cases where the replacement is obvious.
2. Implement other cases as a hand-written analyzer, to avoid false positives.

**Comment by quasilyte** [comment]
@cristaloleg, what do you think?

File: checkers/rules/rules.go (hunk 1 above)
Comment (quasilyte) [reply checkers/rules/rules.go:746]: Maybe this message comes from go-perfguard style warnings.
But there is a difference there: it doesn't report anything if it has a replacement, so it just reports the suggestion template there. If there is no suggestion available, it can only say "don't do that" because it doesn't know how to fix the code.

In my opinion, reporting performance issue without a fix is only useful if it's a big performance issue.

File: checkers/rules/rules.go (hunk 3 above)
Comment (peakle) [reply checkers/rules/rules.go]: +

File: checkers/rules/rules.go (hunk 2 above)
Comment (peakle) [reply checkers/rules/rules.go]: changed to style

File: checkers/rules/rules.go (hunk 1 above)
Comment (peakle) [reply checkers/rules/rules.go]: \+

File: checkers/rules/rules.go [hunk 4]
```diff
@@ -736,3 +736,24 @@ func dynamicFmtString(m dsl.Matcher) {
 		Suggest("errors.New($f($*args))").
 		Report(`use errors.New($f($*args)) or fmt.Errorf("%s", $f($*args)) instead`)
 }
+
+//doc:summary Detects strings.Compare usage
+//doc:tags    style experimental
+//doc:before  strings.Compare(x, y)
+//doc:after   x < y
+func stringsCompare(m dsl.Matcher) {
+	m.Match(`strings.Compare($s1, $s2) == 0`, `0 == strings.Compare($s1, $s2)`).
```
Comment (quasilyte) [inline checkers/rules/rules.go:745]: There is no need to add a yoda-style pattern here.

They should be reported by a separate checker that just checks for `0 == $_` (not exactly, but that's the idea).

Comment (peakle) [reply checkers/rules/rules.go]: \+

**Review by quasilyte — CHANGES_REQUESTED** [review]
Please, cover all cases with tests.
I don't see -1 value case.
Whether possible, all corner cases should be tested.

It's also a good practice to have "fixed" code examples in the negative tests suite.
For example, if linter suggests `>` operator, negative tests suite should contain the suggested code snippet, to be sure that this replacements does not trigger any other warning.

**Review by quasilyte — APPROVED** [review]

**Comment by quasilyte** [comment]
LGTM.
@cristaloleg wanna make a final review? :)

**Review by quasilyte — APPROVED** [review]
